playbooks/roles/web/templates/etc/varnish/default.vcl.j2

```diff
@@ -33,54 +33,110 @@ acl purge {
 # sub vcl_recv #
 ################
 sub vcl_hash {
-  hash_data(req.http.host);
-  hash_data(req.url);
+  # Hash protocol and URL.
   hash_data(req.http.X-Forwarded-Proto);
-  return(hash);
+  hash_data(req.url);
+
+  # Hash host or server IP.
+  if (req.http.host) {
+      hash_data(req.http.host);
+  } else {
+      hash_data(server.ip);
+  }
+
+  # Hash cookies for requests that have them.
+  if (req.http.Cookie) {
+      hash_data(req.http.Cookie);
+  }
+
+  return (hash);
 }
 
 ################
 # sub vcl_recv #
 ################
 sub vcl_recv {
+  # Only deal with "normal" method requests. (http://book.varnish-software.com/3.0/VCL_Basics.html#default-vcl-recv)
+  if (req.request != "GET" &&
+      req.request != "HEAD" &&
+      req.request != "PUT" &&
+      req.request != "POST" &&
+      req.request != "TRACE" &&
+      req.request != "OPTIONS" &&
+      req.request != "PATCH" &&
+      req.request != "DELETE") {
+      /* Non-RFC2616 or CONNECT which is weird. */
+      return (pipe);
+  }
+
+  # Only cache GET or HEAD requests. This makes sure the POST requests are always passed. (http://book.varnish-software.com/3.0/VCL_Basics.html#default-vcl-recv)
+  if (req.request != "GET" && req.request != "HEAD") {
+      return (pass);
+  }
```

Comment:
These two blocks are from [the default config](http://book.varnish-software.com/3.0/VCL_Basics.html#default-vcl-recv). `req.http.X-Forwarded-For` is not included as it need to be handled per case.